### PR TITLE
[CPU] Improve first token latency for 4bit compressed models

### PR DESCRIPTION
**Pull request by dmitry-gorokhov**
### Details:
 - PR improves "first token" latency for 4bit compressed LLMs on AVX2/AVX512 HW targets. 
 - OneDNN PR: https://github.com/openvinotoolkit/oneDNN/pull/221
### Tickets:
 - *[CVS-125052](https://jira.devtools.intel.com/browse/CVS-125052)*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@v-Golubev @xuchen-intel Guys, could you please review. 
Chen, I would ask you to mostly focus on OneDNN changes.

**Review by xuchen-intel — APPROVED** [review]
The OneDNN changes LGTM, with two minor comments.

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp [hunk 1]
```diff
@@ -162,11 +186,14 @@ class MatmulWeightsDecompression : public testing::WithParamInterface<MatmulWeig
         auto scaleshift_const_shape = scaleshift_target_shape;
         if (reshape_on_decompression_constant)
             scaleshift_const_shape.erase(std::remove(scaleshift_const_shape.begin(), scaleshift_const_shape.end(), 1), scaleshift_const_shape.end());
-        if (add_subtract) {
-            auto shift_const = ngraph::builder::makeConstant<uint8_t>(weights_precision, scaleshift_const_shape, {}, true, 7);
+        if (decompression_subtract_type != DecompressionSubtractType::empty) {
+            auto subtract_shape = decompression_subtract_type == DecompressionSubtractType::full ? scaleshift_const_shape : Shape({});
+            auto shift_const = ngraph::builder::makeConstant<uint8_t>(weights_precision, subtract_shape, {}, true, 7);
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp:191]: Should we create `shift_const` in the same way like `weights`?
```suggestion
            auto shift_const = ngraph::builder::makeConstant<int8_t>(weights_precision, subtract_shape, {}, true, up_to);
```

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp [hunk 2]
```diff
@@ -162,11 +186,14 @@ class MatmulWeightsDecompression : public testing::WithParamInterface<MatmulWeig
...
             std::shared_ptr<ov::Node> shift_convert = std::make_shared<ov::op::v0::Convert>(shift_const, decompression_precision);
             if (reshape_on_decompression_constant) {
-                auto shift_reshape_const = ov::opset10::Constant::create(ov::element::i32, {scaleshift_target_shape.size()}, scaleshift_target_shape);
+                auto subtract_target_shape = decompression_subtract_type == DecompressionSubtractType::full
+                    ? scaleshift_target_shape : ov::Shape(std::vector<size_t>(scaleshift_const_shape.size(), 1));
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp:195]: Minor
```suggestion
                    ? scaleshift_target_shape : ov::Shape(scaleshift_const_shape.size(), 1);
```

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp [hunk 3]
```diff
@@ -360,7 +387,7 @@ INSTANTIATE_TEST_SUITE_P(smoke_MatMulCompressedWeights_amx,
 const std::vector<ShapeParams> input_shapes_corner_cases_basic = {
     {{{-1, -1, -1}, {{1, 4, 16}}}, {1, 16, 32}},
     {{{-1, -1, -1}, {{1, 4, 16}}}, {16, 32}},
-    {{{-1, -1, -1}, {{1, 4, 16}}}, {16, 32}, 4ul},
+    {{{-1, -1, -1}, {{1, 5, 16}}}, {16, 32}, 4ul},
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp:390]: I'm just wondering why we changed this test case?

File: src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_fc_reshape_to_weights.cpp [hunk 4]
```diff
@@ -73,7 +73,8 @@ ov::intel_cpu::MoveFCReshapeToWeights::MoveFCReshapeToWeights() {
                 return false;
 
             const auto comparison_start_pos = expected_shape.size() - node_shape.size();
-            return std::equal(node_shape.begin(), node_shape.end(), expected_shape.begin() + comparison_start_pos);
+            return std::equal(node_shape.begin(), node_shape.end(), expected_shape.begin() + comparison_start_pos) ||
+                   std::all_of(node_shape.cbegin(), node_shape.cend(), [](int dim) { return dim == 1; });
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_fc_reshape_to_weights.cpp:77]: Can we cover the updated condition with [unit tests](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/tests/unit/transformations/move_fc_reshape_to_weights.cpp)?

**Review by v-Golubev — APPROVED** [review]
No critical comments from my side

File: src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_fc_reshape_to_weights.cpp (hunk 4 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_fc_reshape_to_weights.cpp:77]: Agree, covered

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp (hunk 3 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp:390]: Just to trigger new decompression algo which has condition on that dimension

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp (hunk 2 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp]: Refactored

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp (hunk 1 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/tests/functional/subgraph_tests/src/matmul_weights_decompression.cpp]: Good point. Added

**Pull request merged**